# Exploratory Drug Label Analysis, OpenFDA

Now that we've found what seems to be a pretty interesting dataset via OpenFDA, lets take a longer look and see what's inside.

In [1]:
import os
import pprint
import json
import lxml.etree as etree
import pandas as pd
from pandas.io.json import json_normalize
pprint = pprint.pprint #Set up pretty print

In [2]:
data = json.load(open('../test_data/drug-label-0001-of-0007.json'))
print(data.keys())
data['meta']

dict_keys(['meta', 'results'])


{'last_updated': '2018-12-14',
 'terms': 'https://open.fda.gov/terms/',
 'results': {'skip': 0, 'total': 131198, 'limit': 20000},
 'license': 'https://open.fda.gov/license/',
 'disclaimer': 'Do not rely on openFDA to make decisions regarding medical care. While we make every effort to ensure that data is accurate, you should assume all results are unvalidated. We may limit or otherwise restrict your access to the API in line with our Terms of Service.'}

In [7]:
#df = pd.DataFrame(data['results'])
df = json_normalize(data['results'])
pprint(df.columns)
with pd.option_context('display.max_columns', 160):
    display(df.head())

Index(['abuse', 'accessories', 'active_ingredient', 'active_ingredient_table',
       'adverse_reactions', 'adverse_reactions_table',
       'animal_pharmacology_and_or_toxicology', 'ask_doctor',
       'ask_doctor_or_pharmacist', 'ask_doctor_or_pharmacist_table',
       ...
       'use_in_specific_populations_table', 'user_safety_warnings', 'version',
       'veterinary_indications', 'warnings', 'warnings_and_cautions',
       'warnings_and_cautions_table', 'warnings_table', 'when_using',
       'when_using_table'],
      dtype='object', length=153)


abuse accessories                                  active_ingredient  \
0   NaN         NaN  [HPUS Active Ingredients Each dose contains eq...   
1   NaN         NaN          [Active ingredient Colloidal oatmeal 43%]   
2   NaN         NaN                                                NaN   
3   NaN         NaN  [Active ingredient (in each extended-release t...   
4   NaN         NaN                                                NaN   

  active_ingredient_table                                  adverse_reactions  \
0                     NaN                                                NaN   
1                     NaN                                                NaN   
2                     NaN  [6 ADVERSE REACTIONS Use of levocetirizine dih...   
3                     NaN                                                NaN   
4                     NaN  [6 ADVERSE REACTIONS The most common adverse r...   

                             adverse_reactions_table  \
0                                                NaN   
1                                                NaN   
2  [<table ID="_RefID0EWPAC" width="75%"> <captio...   
3                                                NaN   
4  [<table ID="_RefID0EOOAG" width="100%"> <capti...   

               animal_pharmacology_and_or_toxicology  \
0                                                NaN   
1                                                NaN   
2  [13.2 Animal Toxicology Reproductive Toxicolog...   
3                                                NaN   
4                                                NaN   

                                          ask_doctor ask_doctor_or_pharmacist  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3  [Ask a doctor before use if you have •persiste...                      NaN   
4                                                NaN                      NaN   

  ask_doctor_or_pharmacist_table ask_doctor_table  \
0                            NaN              NaN   
1                            NaN              NaN   
2                            NaN              NaN   
3                            NaN              NaN   
4                            NaN              NaN   

  assembly_or_installation_instructions  \
0                                   NaN   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4                                   NaN   

                                       boxed_warning boxed_warning_table  \
0                                                NaN                 NaN   
1                                                NaN                 NaN   
2                                                NaN                 NaN   
3                                                NaN                 NaN   
4  [WARNING: FETAL TOXICITY • When pregnancy is d...                 NaN   

  calibration_instructions  \
0                      NaN   
1                      NaN   
2                      NaN   
3                      NaN   
4                      NaN   

  carcinogenesis_and_mutagenesis_and_impairment_of_fertility  \
0                                                NaN           
1                                                NaN           
2  [13.1 Carcinogenesis, Mutagenesis, Impairment ...           
3                                                NaN           
4  [13.1 Carcinogenesis, Mutagenesis, Impairment ...           

  carcinogenesis_and_mutagenesis_and_impairment_of_fertility_table cleaning  \
0                                                NaN                    NaN   
1                                                NaN                    NaN   
2                                                NaN                    NaN  

I think we might be able to work with this now.  We've used `json_normalize` to flatten the semi-structured format into something tabular.  There are 153 column labels, but we'll want to drop a bunch and just use a subset for now. I want to just look at the 'openfda' subset of columns, as it looks like there's some useful info there (https://open.fda.gov/apis/openfda-fields/).

In [43]:
df_openfda = df.filter(regex='openfda*')
print(df_openfda.columns, len(df_openfda.columns))
with pd.option_context('display.max_columns', 25):
    display(df_openfda.head())

Index(['openfda.application_number', 'openfda.brand_name',
       'openfda.generic_name', 'openfda.is_original_packager',
       'openfda.manufacturer_name', 'openfda.nui',
       'openfda.original_packager_product_ndc', 'openfda.package_ndc',
       'openfda.pharm_class_cs', 'openfda.pharm_class_epc',
       'openfda.pharm_class_moa', 'openfda.pharm_class_pe',
       'openfda.product_ndc', 'openfda.product_type', 'openfda.route',
       'openfda.rxcui', 'openfda.spl_id', 'openfda.spl_set_id',
       'openfda.substance_name', 'openfda.unii', 'openfda.upc'],
      dtype='object') 21


openfda.application_number  \
0                        NaN   
1                  [part347]   
2               [ANDA203646]   
3               [ANDA091009]   
4               [ANDA076549]   

                                  openfda.brand_name  \
0                                      [BrightSpark]   
1  [Aveeno Baby Eczema Therapy Soothing Bath Trea...   
2                   [levocetirizine dihydrochloride]   
3                  [Good Neighbor Pharmacy Mucus ER]   
4                                         [Ramipril]   

                                openfda.generic_name  \
0  [ARG NIT, ARSENICUM IOD, HYOSCYAMUS, VERATRUM ...   
1                                          [OATMEAL]   
2                   [LEVOCETIRIZINE DIHYDROCHLORIDE]   
3                                      [GUAIFENESIN]   
4                                         [RAMIPRIL]   

  openfda.is_original_packager          openfda.manufacturer_name  \
0                       [True]               [Silver Star Brands]   
1                       [True]  [Johnson & Johnson Consumer Inc.]   
2                          NaN                 [Proficient Rx LP]   
3                       [True]               [Amerisource Bergen]   
4                          NaN                 [Proficient Rx LP]   

                                         openfda.nui  \
0                                                NaN   
1  [N0000185016, N0000175629, N0000008246, N00001...   
2                                                NaN   
3                                                NaN   
4                         [N0000175562, N0000000181]   

  openfda.original_packager_product_ndc  \
0                                   NaN   
1                                   NaN   
2                           [42543-712]   
3                                   NaN   
4                           [16252-572]   

                          openfda.package_ndc  \
0                              [68703-252-02]   
1                              [69968-0199-5]   
2  [63187-663-60, 63187-663-90, 63187-663-30]   
3                              [46122-351-74]   
4  [63187-675-30, 63187-675-90, 63187-675-60]   

                              openfda.pharm_class_cs  \
0                                                NaN   
1  [Grain Proteins [Chemical/Ingredient], Dietary...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                             openfda.pharm_class_epc  \
0                                                NaN   
1  [Non-Standardized Food Allergenic Extract [EPC...   
2                                                NaN   
3                                                NaN   
4    [Angiotensin Converting Enzyme Inhibitor [EPC]]   

                            openfda.pharm_class_moa  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4  [Angiotensin-converting Enzyme Inhibitors [MoA]]   

                              openfda.pharm_class_pe openfda.product_ndc  \
0                                                NaN         [68703-252]   
1  [Cell-mediated Immunity [PE], Increased Histam...        [69968-0199]   
2                                                NaN         [63187-663]   
3                                                NaN         [46122-351]   
4                                                NaN         [63187-675]   

        openfda.product_type openfda.route openfda.rxcui  \
0           [HUMAN OTC DRUG]        [ORAL]           NaN   
1           [HUMAN OTC DRUG]     [TOPICAL]           NaN   
2  [HUMAN PRESCRIPTION DRUG]        [ORAL]      [855172]   
3           [HUMAN OTC DRUG]        [ORAL]      [310621]   
4  [HUMAN PRESCRIPTION DRUG]        [ORAL]      [198189]   

      

Ok, looking good.  Are these brackets strings that were parsed wrong or actual list objects?

In [19]:
print(df_openfda.iloc[0][1])
type(df_openfda.iloc[0][1])

['BrightSpark']


list

That answers that.  I don't want a bunch of lists for anything that's a single item, so lets unpack all lists of length=1.  Lets also strip 'openfda' off of our columns.

In [44]:
df_openfda = df_openfda.applymap(lambda x: x[0] if type(x) is list and len(x)==1 else x)
df_openfda.columns = df_openfda.columns.str.replace('openfda.','')

In [45]:
df_openfda.head()

application_number                                         brand_name  \
0                NaN                                        BrightSpark   
1            part347  Aveeno Baby Eczema Therapy Soothing Bath Treat...   
2         ANDA203646                     levocetirizine dihydrochloride   
3         ANDA091009                    Good Neighbor Pharmacy Mucus ER   
4         ANDA076549                                           Ramipril   

                                       generic_name is_original_packager  \
0  ARG NIT, ARSENICUM IOD, HYOSCYAMUS, VERATRUM ALB                 True   
1                                           OATMEAL                 True   
2                    LEVOCETIRIZINE DIHYDROCHLORIDE                  NaN   
3                                       GUAIFENESIN                 True   
4                                          RAMIPRIL                  NaN   

                 manufacturer_name  \
0               Silver Star Brands   
1  Johnson & Johnson Consumer Inc.   
2                 Proficient Rx LP   
3               Amerisource Bergen   
4                 Proficient Rx LP   

                                                 nui  \
0                                                NaN   
1  [N0000185016, N0000175629, N0000008246, N00001...   
2                                                NaN   
3                                                NaN   
4                         [N0000175562, N0000000181]   

  original_packager_product_ndc                                 package_ndc  \
0                           NaN                                68703-252-02   
1                           NaN                                69968-0199-5   
2                     42543-712  [63187-663-60, 63187-663-90, 63187-663-30]   
3                           NaN                                46122-351-74   
4                     16252-572  [63187-675-30, 63187-675-90, 63187-675-60]   

                                      pharm_class_cs  \
0                                                NaN   
1  [Grain Proteins [Chemical/Ingredient], Dietary...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                     pharm_class_epc      ...        \
0                                                NaN      ...         
1  [Non-Standardized Food Allergenic Extract [EPC...      ...         
2                                                NaN      ...         
3                                                NaN      ...         
4      Angiotensin Converting Enzyme Inhibitor [EPC]      ...         

                                      pharm_class_pe product_ndc  \
0                                                NaN   68703-252   
1  [Cell-mediated Immunity [PE], Increased Histam...  69968-0199   
2                                                NaN   63187-663   
3                                                NaN   46122-351   
4                                                NaN   63187-675   

              product_type    route   rxcui  \
0           HUMAN OTC DRUG     ORAL     NaN   
1           HUMAN OTC DRUG  TOPICAL     NaN   
2  HUMAN PRESCRIPTION DRUG     ORAL  855172   
3           HUMAN OTC DRUG     ORAL  310621   
4  HUMAN PRESCRIPTION DRUG     ORAL  198189   

                                 spl_id                            spl_set_id  \
0  2eb93d7c-3310-4b93-e054-00144ff88e88  2eb93d7c-330f-4b93-e054-00144ff88e88   
1  ae390f8a-fdf1-4711-9943-e1a46e2ef0e8  7f689b72-ccfb-4ae0-9747-1490593998ca   
2  b84ccb27-9c52-43b9-9a46-8f52a571045e  b84ccb27-9c52-43b9-9a46-8f52a571045e   
3  1e99f1a3-95bc-4108-8395-430a32c4f292  1e99f1a3-95bc-4108-8395-430a32c4f292   
4  4146cfc6-2b02-4731-a0f1-8c445bafea55  4146cfc6-2b02-4731-a0f1-8c445bafea55   

                                      substance_name  \
0  [VERATRUM ALBUM ROOT, HYOSCYAMUS NIGER, SILVER...  

Excellent.  I want to reindex on something unique as well, ideally an identifier for the drug.  We should figure out some basics about each column before we use it as an index (or do anything with it, really) 

In [46]:
for col in df_openfda:
    try:
        display(df_openfda[col].describe())
    except TypeError:
        print(col, 'cannot be described')

count       15942
unique       2088
top       part352
freq         3496
Name: application_number, dtype: object

count      20000
unique     14073
top       Oxygen
freq         127
Name: brand_name, dtype: object

count       20000
unique       4877
top       ALCOHOL
freq          650
Name: generic_name, dtype: object

count     16920
unique        1
top        True
freq      16920
Name: is_original_packager, dtype: object

count                                           20000
unique                                           3601
top       Rxhomeo Private Limited d.b.a. Rxhomeo, Inc
freq                                              565
Name: manufacturer_name, dtype: object

nui cannot be described
original_packager_product_ndc cannot be described
package_ndc cannot be described
pharm_class_cs cannot be described
pharm_class_epc cannot be described
pharm_class_moa cannot be described
pharm_class_pe cannot be described
product_ndc cannot be described


count              20000
unique                 2
top       HUMAN OTC DRUG
freq               15851
Name: product_type, dtype: object

route cannot be described
rxcui cannot be described


count                                    20000
unique                                   20000
top       9c8aa994-6715-471d-9fc7-b88d21eec369
freq                                         1
Name: spl_id, dtype: object

count                                    20000
unique                                   20000
top       3f78386f-42f1-22a9-1f27-349381dad0b0
freq                                         1
Name: spl_set_id, dtype: object

substance_name cannot be described
unii cannot be described
upc cannot be described


Only two columns are completely unique and do not contain lists: spl_id and spl_set_id.  Brand Names is not unique, but it's close.  Apparently a lot of companies make 'Oxygen'.  Good, I'm a firm believer that oxygen is important.  Let's take a look.

In [50]:
df_openfda[df_openfda.brand_name == 'Oxygen']

application_number brand_name generic_name is_original_packager  \
1275                 NaN     Oxygen       OXYGEN                 True   
1411                 NaN     Oxygen       OXYGEN                 True   
2173                 NaN     Oxygen       OXYGEN                 True   
2540           NDA205865     Oxygen       OXYGEN                 True   
2590                 NaN     Oxygen       OXYGEN                 True   
2789                 NaN     Oxygen       OXYGEN                 True   
3502                 NaN     Oxygen       OXYGEN                 True   
3848           NDA205849     Oxygen       OXYGEN                 True   
4210                 NaN     Oxygen       OXYGEN                 True   
4493                 NaN     Oxygen       OXYGEN                 True   
4728                 NaN     Oxygen       OXYGEN                 True   
5175                 NaN     Oxygen       OXYGEN                 True   
5222                 NaN     Oxygen       OXYGEN                 True   
5321                 NaN     Oxygen       OXYGEN                 True   
5993                 NaN     Oxygen       OXYGEN                 True   
6106                 NaN     Oxygen       OXYGEN                 True   
6140                 NaN     Oxygen       OXYGEN                 True   
6648                 NaN     Oxygen       OXYGEN                 True   
6678                 NaN     Oxygen       OXYGEN                 True   
7341                 NaN     Oxygen       OXYGEN                 True   
7346                 NaN     Oxygen       OXYGEN                 True   
7381                 NaN     Oxygen       OXYGEN                 True   
7441                 NaN     Oxygen       OXYGEN                 True   
7466                 NaN     Oxygen       OXYGEN                 True   
7741                 NaN     Oxygen       OXYGEN                 True   
8116                 NaN     Oxygen       OXYGEN                 True   
8437                 NaN     Oxygen       OXYGEN                 True   
9399                 NaN     Oxygen       OXYGEN                 True   
9498                 NaN     Oxygen       OXYGEN                 True   
9499                 NaN     Oxygen       OXYGEN                 True   
...                  ...        ...          ...                  ...   
19074                NaN     Oxygen       OXYGEN                 True   
19078                NaN     Oxygen       OXYGEN                 True   
19166                NaN     Oxygen       OXYGEN                 True   
19169                NaN     Oxygen       OXYGEN                 True   
19180                NaN     Oxygen       OXYGEN                 True   
19261                NaN     Oxygen       OXYGEN                 True   
19274                NaN     Oxygen       OXYGEN                 True   
19318                NaN     Oxygen       OXYGEN                 True   
19376                NaN     Oxygen       OXYGEN                 True   
19377                NaN     Oxygen       OXYGEN                 True   
19410                NaN     Oxygen       OXYGEN                 True   
19414                NaN     Oxygen       OXYGEN                 True   
19470                NaN     Oxygen       OXYGEN                 True   
19473                NaN     Oxygen       OXYGEN                 True   
19514                NaN     Oxygen       OXYGEN                 True   
19516                NaN     Oxygen       OXYGEN                 True   
19520                NaN     Oxygen       OXYGEN                 True   
19581                NaN     Oxygen       OXYGEN                 True   
19617                NaN     Oxygen       OXYGEN                 True   
19619                NaN     Oxygen       OXYGEN                 True   
19683                NaN     Oxygen       OXYGEN                 True   
19714                NaN     Oxygen       OXYGEN                 True   
19727                NaN     Oxygen       OXYGEN                

Ok, what if we multiindex this a bit, just to make things more readable or organized.  How about, generic->brand?

In [54]:
print(df_openfda.generic_name.isnull().any())
print(df_openfda.brand_name.isnull().any())
print(df_openfda.spl_id.isnull().any())

False
False
False


In [69]:
df_openfda.set_index(['generic_name','brand_name']).sort_index()

application_number  \
generic_name                                       brand_name                                             
.ALPHA.-KETOGLUTARIC ACID - AMBERGRIS - ANAMIRT... GUNA-DIZZY                                       NaN   
.GAMMA.-AMINOBUTYRIC ACID                          Gongjinhyang Qi and Jin Eye                      NaN   
1.85% HYDROCORTISONE ACETATE - 1.15% PRAMOXINE ... PROCORT                                          NaN   
38 BIOTECH LABORATORY LLC                          Chitoprex 101                                    NaN   
5-HYDROXYTRYPTOPHAN                                5-HTP Phenolic                                   NaN   
50 REMEDY KIT                                      Homeopathic Remedy Kit 30c                       NaN   
62% ETHYL ALCOHOL                                  Hand Sanitizer Orange with aloe             part333A   
                                                   hand sanitizer orange with aloe             part333A   
ABACAVIR SULFATE                                   Abacavir Sulfate                          ANDA091294   
                                                   ZIAGEN                                     NDA020977   
ABACAVIR SULFATE AND LAMIVUDINE                    EPZICOM                                    NDA021652   
ABACAVIR SULFATE, LAMIVUDINE, AND ZIDOVUDINE       TRIZIVIR                                   NDA021205   
ABELMOSCHUS                                        ABELMOSCHUS                                      NaN   
ABIES CANADENSIS                                   ABIES CANADENSIS                                 NaN   
                                                   ABIES CANADENSIS                                 NaN   
ABIES CANADENSIS, ALLIUM CEPA, ALLIUM SATIVUM, ... Kids Candida Relief                              NaN   
ABIES NIGRA                                        ABIES NIGRA                                      NaN   
                                                   ABIES NIGRA                                      NaN   
ABIES NIGRA, ANTIMONIUM CRUDUM, ASAFOETIDA, BAP... Yeast Freee                                      NaN   
ABIES NIGRA, ARSENICUM ALBUM, CONDURANGO, CONIU... Heartburn Relief                                 NaN   
ABIES NIGRA, CONDURANGO, CONIUM MACULATUM, IRIS... HeartburnCare                                    NaN   
ABROTANUM                                          ABROTANUM                                        NaN   
                                                   ABROTANUM                                        NaN   
                                                   ABROTANUM                                        NaN   
ABROTANUM (ARTEMISIA ABROTANUM), ANACARDIUM ORI... Ultra Advantage                                  NaN   
ABROTANUM, ADRENALINUM, ADRENOCORTICOTROPHIN, A... Anti Aging and Wrinkles For Women                NaN   
ABROTANUM, AESCULUS HIPP., ALLIUM SATIVUM, ARSE... Parasites                                        NaN   
ABROTANUM, ANACARDIUM ORIENTALE, ARSENICUM ALBU... Anti-Aging and Wrinkles                          NaN   
ABROTANUM, ANACARDIUM ORIENTALE, BARYTA CARBONI... SWL ULTRA ADVANTAGE                              NaN   
ABROTANUM, ANACARDIUM ORIENTALE, BARYTA MURIATI... ADVANTAGE                                        NaN   
...                                                                                                 ...   
ZOLPIDEM TARTRATE                                  ZOLPIDEM TARTRATE                         ANDA077903   
                                                   ZOLPIDEM TARTRATE                         ANDA077903   
                                                   ZOLPIDEM TARTRATE                         ANDA078426   
                                                   ZOLPIDEM TARTRATE                         ANDA077903   
                                                   ZOLPIDEM TARTRATE                         ANDA077903   
                   

It's something, at least.  I'm not sure we want to make the kind of distinctions like 62% Ethyl Alcohol vs... how many ethyls do we have?

In [79]:
df_openfda.set_index(['generic_name','brand_name']).sort_index().filter(regex=r'\bETHYL\b', axis=0)

application_number  \
generic_name                           brand_name                                                              
62% ETHYL ALCOHOL                      Hand Sanitizer Orange with aloe                              part333A   
                                       hand sanitizer orange with aloe                              part333A   
ALCOHOL                                ETHYL ALCOHOL                                                 part333   
                                       ETHYL ALCOHOL                                                 part333   
                                       ETHYL ALCOHOL                                                 part333   
                                       ETHYL ALCOHOL                                                 part333   
                                       ETHYL ALCOHOL                                                 part333   
CHLORHEXIDINE GLUCONATE, ETHYL ALCOHOL Purmist                                                       part333   
                                       Purmist                                                       part333   
ETHYL ALCOHOL                          2-Pack Hand Sanitizer                                        part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI BACTERIAL HAND SANITIZER                                part333E   
                                       ANTI-AGING HAND SANITIZER                                    part333E   
                                       ANTIBACTERIAL HAND SANITIZER BASIL AND GREEN TEA             part333E   
                                       ANTIBACTERIAL SANITIZER                                      part333E   
                                       AURORA ANTIBACTERIAL SANITIZER SUMMER COTTON BR...           part333E   
                                       AURORA ANTIBACTERIAL SANITIZER SWEET LYCHEE BERRY            part333E   
                                       Advanced                                                     part333A   
                                       Advanced                                                     part333A   
                                       AmeriDerm Foam Alcohol Hand Sanitizer                        part333E   
...                                                                                                      ...   
                                       anti-bacterial hand sanitizer                                part333E   
                                       anti-bacterial hand sanitizer                                part333E   
                                       anti-bacterial hand sanitizer                                part333E   
                                       bubble gum         

We'll have to think on this a bit.  But we're getting somewhere and it's been pretty easy to manipulate the data.